File: javascript/packages/highlighter/src/highlighter.ts
import { readFileSync } from "fs"

import { SyntaxRenderer } from "./syntax-renderer.js"
import { DiagnosticRenderer } from "./diagnostic-renderer.js"
import { FileRenderer } from "./file-renderer.js"
import { InitializationManager } from "./initialization-manager.js"
import { InlineDiagnosticRenderer } from "./inline-diagnostic-renderer.js"
import { DiffRenderer } from "./diff-renderer.js"
import { LineWrapper } from "./line-wrapper.js"
import { resolveTheme } from "./themes.js"

import type { HerbBackend, Diagnostic } from "@herb-tools/core"
import type { ThemeInput } from "./themes.js"
import type { DiffRenderOptions } from "./diff-renderer.js"
import type { DiffHunk } from "./diff-computer.js"

export interface HighlightOptions {
  diagnostics?: Diagnostic[]
  splitDiagnostics?: boolean

  contextLines?: number
  focusLine?: number
  showLineNumbers?: boolean
  wrapLines?: boolean
  maxWidth?: number
  truncateLines?: boolean
  codeUrlBuilder?: (code: string) => string
  fileUrlBuilder?: (path: string, diagnostic: Diagnostic) => string
  suffixBuilder?: (diagnostic: Diagnostic) => string | undefined
}

export interface HighlightDiagnosticOptions {
  contextLines?: number
  showLineNumbers?: boolean
  optimizeHighlighting?: boolean
  wrapLines?: boolean
  maxWidth?: number
  truncateLines?: boolean
  codeUrl?: string
  fileUrl?: string
  suffix?: string
}

export class Highlighter {
  private syntaxRenderer: SyntaxRenderer
  private diagnosticRenderer: DiagnosticRenderer
  private fileRenderer: FileRenderer
  private initManager: InitializationManager
  private inlineDiagnosticRenderer: InlineDiagnosticRenderer
  private diffRenderer: DiffRenderer

  constructor(theme: ThemeInput = "onedark", herb?: HerbBackend) {
    const colors = resolveTheme(theme)
    this.syntaxRenderer = new SyntaxRenderer(colors, herb)
    this.diagnosticRenderer = new DiagnosticRenderer(this.syntaxRenderer)
    this.fileRenderer = new FileRenderer(this.syntaxRenderer)
    this.initManager = new InitializationManager(herb)
    this.inlineDiagnosticRenderer = new InlineDiagnosticRenderer(this.syntaxRenderer)
    this.diffRenderer = new DiffRenderer(this.syntaxRenderer, colors)
  }

  /**
   * Initialize the highlighter with the Herb backend
   * This must be called before using highlight() or highlightDiagnostic()
   */
  async initialize(): Promise<void> {
    await this.initManager.initialize()
    await this.syntaxRenderer.initialize()
  }

  /**
   * Check if the highlighter has been initialized
   */
  get initialized(): boolean {
    return this.initManager.initialized
  }

  requireInitialized(): void {
    this.initManager.requireInitialized()
  }

  /**
   * Main highlighting method with flexible rendering options
   * @param path - File path for annotation (display only, not used for reading)
   * @param content - The content to highlight
   * @param options - Configuration options
   *   - diagnostics: Array of diagnostics to display inline or split
   *   - splitDiagnostics: When true with diagnostics, render each diagnostic individually
   *   - contextLines: Number of context lines around focus/diagnostics
   *   - focusLine: Line number to focus on (shows only that line with dimmed context)
   *   - showLineNumbers: Whether to show line numbers (default: true)
   * @returns The highlighted content with optional diagnostics or focused view
   */
  highlight(
    path: string,
    content: string,
    options: HighlightOptions = {},
  ): string {
    this.requireInitialized()

    const {
      diagnostics = [],
      splitDiagnostics = false,
      contextLines = 0,
      focusLine,
      showLineNumbers = true,
      wrapLines = true,
      maxWidth = LineWrapper.getTerminalWidth(),
      truncateLines = false,
      codeUrlBuilder,
      fileUrlBuilder,
      suffixBuilder,
    } = options

    // Case 1: Split diagnostics - render each diagnostic individually
    if (diagnostics.length > 0 && splitDiagnostics) {
      const results: string[] = []
      for (let i = 0; i < diagnostics.length; i++) {
        const diagnostic = diagnostics[i]
        const codeUrl = codeUrlBuilder && diagnostic.code ? codeUrlBuilder(diagnostic.code) : undefined
        const fileUrl = fileUrlBuilder ? fileUrlBuilder(path, diagnostic) : undefined
        const suffix = suffixBuilder ? suffixBuilder(diagnostic) : undefined
        const result = this.highlightDiagnostic(path, diagnostic, content, {
          contextLines,
          showLineNumbers,
          wrapLines,
          maxWidth,
          truncateLines,
          codeUrl,
          fileUrl,
          suffix,
        })

        results.push(result)

        if (i < diagnostics.length - 1) {
          const width = LineWrapper.getTerminalWidth()
          const progressText = `[${i + 1}/${diagnostics.length}]`
          const rightPadding = 16
          const separatorLength = Math.max(0, width - progressText.length - 1 - rightPadding)
          const separator = '⎯'
          const leftSeparator = separator.repeat(separatorLength)
          const rightSeparator = separator.repeat(4)
          const progress = progressText

          results.push(`${leftSeparator}  ${progress} ${rightSeparator}`)
        }
      }
      return results.join("\n\n")
    }

    // Case 2: Inline diagnostics - show whole file with diagnostics inline
    if (diagnostics.length > 0) {
      return this.inlineDiagnosticRenderer.render(
        path,
        content,
        diagnostics,
        showLineNumbers,
        wrapLines,
        maxWidth,
        truncateLines,
        codeUrlBuilder,
      )
    }

    // Case 3: Focus line - show only specific line with context
    if (focusLine) {
      return this.fileRenderer.renderWithFocusLine(
        path,
        content,
        focusLine,
        contextLines,
        showLineNumbers,
        maxWidth,
        wrapLines,
        truncateLines,
      )
    }

    // Case 4: Default - just highlight the whole file
    if (showLineNumbers) {
      return this.fileRenderer.renderWithLineNumbers(path, content, wrapLines, maxWidth, truncateLines)
    } else {
      return this.fileRenderer.renderPlain(content, maxWidth, wrapLines, truncateLines)
    }
  }

  /**
   * Render a single diagnostic with context lines and syntax highlighting
   * @param path - The file path to display in the diagnostic (display only)
   * @param diagnostic - The diagnostic message to render
   * @param content - The content to highlight and render
   * @param options - Optional configuration
   * @returns The rendered diagnostic output with syntax highlighting
   */
  highlightDiagnostic(
    path: string,
    diagnostic: Diagnostic,
    content: string,
    options: HighlightDiagnosticOptions = {},
  ): string {
    this.requireInitialized()

    return this.diagnosticRenderer.renderSingle(
      path,
      diagnostic,
      content,
      options,
    )
  }

  /**
   * Render the change between two sources as a syntax-highlighted diff
   * @param path - File path shown above the diff (display only)
   * @param original - The source before the change
   * @param modified - The source after the change
   * @param options - Optional configuration
   * @returns The rendered diff, or an empty string when the sources are identical
   */
  highlightDiff(
    path: string,
    original: string,
    modified: string,
    options: DiffRenderOptions = {},
  ): string {
    this.requireInitialized()

    return this.diffRenderer.render(path, original, modified, options)
  }

  /**
   * Render pre-computed diff hunks
   * @param path - File path shown above the diff (display only)
   * @param hunks - The hunks to render
   * @param options - Optional configuration
   * @returns The rendered diff, or an empty string when there are no hunks
   */
  highlightDiffHunks(
    path: string,
    hunks: DiffHunk[],
    options: DiffRenderOptions = {},
  ): string {
    this.requireInitialized()

    return this.diffRenderer.renderFromHunks(path, hunks, options)
  }

  // File reading wrapper functions

  /**
   * Convenience method that reads a file and highlights it
   * @param filePath - Path to the file to read and highlight
   * @param options - Configuration options
   * @returns The highlighted file content with optional diagnostics
   */
  highlightFileFromPath(
    filePath: string,
    options: HighlightOptions = {},
  ): string {
    this.requireInitialized()

    return this.highlight(filePath, readFile(filePath), options)
  }

  /**
   * Convenience method that reads a file and renders a diagnostic
   * @param filePath - Path to the file to read
   * @param diagnostic - The diagnostic message to render
   * @param options - Optional configuration
   * @returns The highlighted diagnostic output
   */
  highlightDiagnosticFromPath(
    filePath: string,
    diagnostic: Diagnostic,
    options: HighlightDiagnosticOptions = {},
  ): string {
    this.requireInitialized()

    return this.highlightDiagnostic(filePath, diagnostic, readFile(filePath), options)
  }
}

function readFile(filePath: string): string {
  try {
    return readFileSync(filePath, "utf8")
  } catch (error) {
    throw new Error(`Failed to read file ${filePath}: ${error instanceof Error ? error.message : String(error)}`)
  }
}

/**
 * Convenience function to highlight content with a specific theme
 * @param content - The content to highlight
 * @param theme - The theme to use (defaults to "onedark")
 * @param options - Additional highlighting options
 * @returns The highlighted content
 */
export async function highlightContent(
  content: string,
  theme: ThemeInput = "onedark",
  options: HighlightOptions = {}
): Promise<string> {
  const highlighter = new Highlighter(theme)
  await highlighter.initialize()
  return highlighter.highlight("", content, options)
}

/**
 * Convenience function to highlight a file with a specific theme
 * @param filePath - The path to the file to highlight
 * @param theme - The theme to use (defaults to "onedark")
 * @param options - Additional highlighting options
 * @returns The highlighted file content
 */
export async function highlightFile(
  filePath: string,
  theme: ThemeInput = "onedark",
  options: HighlightOptions = {}
): Promise<string> {
  const highlighter = new Highlighter(theme)
  await highlighter.initialize()
  return highlighter.highlightFileFromPath(filePath, options)
}

File: rust/herb-highlighter/src/highlighter.rs
use std::sync::Arc;

use crate::diagnostic::Diagnostic;
use crate::diagnostic_renderer::{DiagnosticRenderOptions, DiagnosticRenderer};
use crate::diff_computer::DiffHunk;
use crate::diff_renderer::{DiffRenderOptions, DiffRenderer};
use crate::error::HighlightError;
use crate::file_renderer::{FileRenderer, RenderOptions};
use crate::herb_backend::{Herb, HerbBackend};
use crate::inline_diagnostic_renderer::{CodeUrlBuilder, InlineDiagnosticRenderer};
use crate::line_wrapper::LineWrapper;
use crate::syntax_renderer::SyntaxRenderer;
use crate::themes::{resolve_theme, ColorScheme, DEFAULT_THEME};

pub type FileUrlBuilder<'a> = &'a dyn Fn(&str, &Diagnostic) -> String;

pub type SuffixBuilder<'a> = &'a dyn Fn(&Diagnostic) -> Option<String>;

pub struct HighlightOptions<'a> {
  pub diagnostics: &'a [Diagnostic],
  pub split_diagnostics: bool,

  pub context_lines: usize,
  pub focus_line: Option<usize>,
  pub show_line_numbers: bool,
  pub wrap_lines: bool,
  pub max_width: Option<usize>,
  pub truncate_lines: bool,
  pub code_url_builder: Option<CodeUrlBuilder<'a>>,
  pub file_url_builder: Option<FileUrlBuilder<'a>>,
  pub suffix_builder: Option<SuffixBuilder<'a>>,
}

impl Default for HighlightOptions<'_> {
  fn default() -> Self {
    Self {
      diagnostics: &[],
      split_diagnostics: false,
      context_lines: 0,
      focus_line: None,
      show_line_numbers: true,
      wrap_lines: true,
      max_width: None,
      truncate_lines: false,
      code_url_builder: None,
      file_url_builder: None,
      suffix_builder: None,
    }
  }
}

pub type HighlightDiagnosticOptions = DiagnosticRenderOptions;

pub struct Highlighter {
  colors: ColorScheme,
  syntax_renderer: SyntaxRenderer,
}

impl Highlighter {
  pub fn new(theme: &str) -> Result<Self, HighlightError> {
    Self::with_backend(theme, Arc::new(Herb))
  }

  pub fn with_backend(theme: &str, herb: Arc<dyn HerbBackend>) -> Result<Self, HighlightError> {
    let colors = resolve_theme(theme)?;

    Ok(Self {
      colors: colors.clone(),
      syntax_renderer: SyntaxRenderer::with_backend(colors, herb),
    })
  }

  pub fn highlight(&self, path: &str, content: &str, options: &HighlightOptions<'_>) -> String {
    let max_width = options.max_width.unwrap_or_else(LineWrapper::get_terminal_width);

    let render_options = RenderOptions {
      show_line_numbers: options.show_line_numbers,
      wrap_lines: options.wrap_lines,
      max_width,
      truncate_lines: options.truncate_lines,
    };

    if !options.diagnostics.is_empty() && options.split_diagnostics {
      let mut results: Vec<String> = Vec::new();

      for (index, diagnostic) in options.diagnostics.iter().enumerate() {
        let code_url = match (options.code_url_builder, &diagnostic.code) {
          (Some(builder), Some(code)) => Some(builder(code)),
          _ => None,
        };

        let file_url = options.file_url_builder.map(|builder| builder(path, diagnostic));
        let suffix = options.suffix_builder.and_then(|builder| builder(diagnostic));

        let result = self.highlight_diagnostic(
          path,
          diagnostic,
          content,
          &HighlightDiagnosticOptions {
            context_lines: options.context_lines,
            show_line_numbers: options.show_line_numbers,
            wrap_lines: options.wrap_lines,
            max_width: Some(max_width),
            truncate_lines: options.truncate_lines,
            code_url,
            file_url,
            suffix,
            ..Default::default()
          },
        );

        results.push(result);

        if index < options.diagnostics.len() - 1 {
          let width = LineWrapper::get_terminal_width();
          let progress_text = format!("[{}/{}]", index + 1, options.diagnostics.len());
          let right_padding = 16;
          let separator_length = width.saturating_sub(progress_text.chars().count() + 1 + right_padding);
          let separator = "⎯";
          let left_separator = separator.repeat(separator_length);
          let right_separator = separator.repeat(4);

          results.push(format!("{left_separator}  {progress_text} {right_separator}"));
        }
      }

      return results.join("\n\n");
    }

    if !options.diagnostics.is_empty() {
      return InlineDiagnosticRenderer::new(&self.syntax_renderer).render(path, content, options.diagnostics, &render_options, options.code_url_builder);
    }

    if let Some(focus_line) = options.focus_line {
      return FileRenderer::new(&self.syntax_renderer).render_with_focus_line(path, content, focus_line, options.context_lines, &render_options);
    }

    let file_renderer = FileRenderer::new(&self.syntax_renderer);

    if options.show_line_numbers {
      file_renderer.render_with_line_numbers(path, content, &render_options)
    } else {
      file_renderer.render_plain(content, &render_options)
    }
  }

  pub fn highlight_diagnostic(&self, path: &str, diagnostic: &Diagnostic, content: &str, options: &HighlightDiagnosticOptions) -> String {
    DiagnosticRenderer::new(&self.syntax_renderer).render_single(path, diagnostic, content, options)
  }

  pub fn highlight_diff(&self, path: &str, original: &str, modified: &str, options: &DiffRenderOptions) -> String {
    DiffRenderer::new(&self.syntax_renderer, self.colors.clone()).render(path, original, modified, options)
  }

  pub fn highlight_diff_hunks(&self, path: &str, hunks: &[DiffHunk], options: &DiffRenderOptions) -> String {
    DiffRenderer::new(&self.syntax_renderer, self.colors.clone()).render_from_hunks(path, hunks, options)
  }

  pub fn highlight_file_from_path(&self, file_path: &str, options: &HighlightOptions<'_>) -> Result<String, HighlightError> {
    let content = read_file(file_path)?;

    Ok(self.highlight(file_path, &content, options))
  }

  pub fn highlight_diagnostic_from_path(
    &self,
    file_path: &str,
    diagnostic: &Diagnostic,
    options: &HighlightDiagnosticOptions,
  ) -> Result<String, HighlightError> {
    let content = read_file(file_path)?;

    Ok(self.highlight_diagnostic(file_path, diagnostic, &content, options))
  }
}

fn read_file(file_path: &str) -> Result<String, HighlightError> {
  std::fs::read_to_string(file_path).map_err(|error| HighlightError::new(format!("Failed to read file {file_path}: {error}")))
}

pub fn highlight_content(content: &str, theme: &str, options: &HighlightOptions<'_>) -> Result<String, HighlightError> {
  Highlighter::new(theme).map(|highlighter| highlighter.highlight("", content, options))
}

pub fn highlight_file(file_path: &str, theme: &str, options: &HighlightOptions<'_>) -> Result<String, HighlightError> {
  Highlighter::new(theme)?.highlight_file_from_path(file_path, options)
}

impl Default for Highlighter {
  fn default() -> Self {
    Self::new(DEFAULT_THEME.as_str()).expect("the default theme is bundled")
  }
}
